infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServerCache.java

```diff
@@ -1,47 +1,66 @@
 package org.corfudb.infrastructure;
 
-import com.github.benmanes.caffeine.cache.Cache;
-import com.github.benmanes.caffeine.cache.CacheWriter;
-import com.github.benmanes.caffeine.cache.Caffeine;
-import com.github.benmanes.caffeine.cache.RemovalCause;
-import com.google.common.annotations.VisibleForTesting;
 import lombok.EqualsAndHashCode;
 import lombok.Getter;
 import lombok.extern.slf4j.Slf4j;
+import javax.annotation.concurrent.NotThreadSafe;
 import org.corfudb.runtime.view.Address;
+import org.corfudb.util.MetricsUtils;
 
-import javax.annotation.Nonnull;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.PriorityQueue;
 import java.util.UUID;
-import java.util.concurrent.atomic.AtomicLong;
 
 /**
  * Sequencer server cache.
  * Contains transaction conflict-resolution data structures.
- * <p>
  * The SequencerServer use its own thread/s. To guarantee correct tx conflict-resolution,
- * the {@link SequencerServerCache#conflictCache} must be updated
+ * the {@link SequencerServerCache#cacheConflictKeys} must be updated
  * along with {@link SequencerServerCache#maxConflictWildcard} at the same time (atomically) to prevent race condition
  * when the conflict stream is already evicted from the cache but `maxConflictWildcard` is not updated yet,
  * which can cause situation when sequencer let the transaction go but the tx has to be cancelled.
- * <p>
+ *
+ * The cache map maps conflict keys (stream id + key) to versions (long), illustrated below:
+ * Conflict Key | ck1 | ck2 | ck3 | ck4
+ * Version | v1 | v1 | v2 | v3
+ * Consider the case where we need to insert a new conflict key (ck), but the cache is full,
+ * we need to evict the oldest conflict keys. In the example above, we can't just evict ck1,
+ * we need to evict all keys that map to v1, so we need to evict ck1 and ck2,
+ * this eviction policy is FIFO on the version number. The simple FIFO opproach in Caffein etc doesn't work here,
+ * as it may evict ck1, but not ck2. Notice that we also can't evict ck3 before the keys for v1,
+ * that's because it will create holes in the resolution window and can lead to incorrect resolutions.
+ *
+ * We use priority queue as a sliding window on the versions, where a version can map to multiple keys,
+ * so we also need to maintain the beginning of the window which is the maxConflictWildcard variable.
+ *
  * SequencerServerCache achieves consistency by using single threaded cache. It's done by following code:
  * `.executor(Runnable::run)`
  */
+@NotThreadSafe
 @Slf4j
 public class SequencerServerCache {
     /**
      * TX conflict-resolution information:
-     * <p>
      * a cache of recent conflict keys and their latest global-log position.
      */
-    private final Cache<ConflictTxStream, Long> conflictCache;
+
+    // As the sequencer cache is used by a single thread, it is safe to use hashmap.
+    private final HashMap<ConflictTxStream, Long> cacheConflictKeys;
```

Comment:
I think you do not need a hashmap here, using a HashSet should be enough because ConflictTxStream already contains the version.

Reply:
HashSet doesn't have get function. Given the key is part of the object, we need to retrieve the address.